File: SimianPie/Examples/phold-opt.py
##############################################################################
# © Copyright 2015-. Triad National Security, LLC. All rights reserved.
#
# This program was produced under U.S. Government contract 89233218CNA000001 for Los Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC for the U.S. Department of Energy/National Nuclear Security Administration.
#
# All rights in the program are reserved by Triad National Security, LLC, and the U.S. Department of Energy/National Nuclear Security Administration. The Government is granted for itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide license in this material to reproduce, prepare derivative works, distribute copies to the public, perform publicly and display publicly, and to permit others to do so.
#
# This is open source software; you can redistribute it and/or modify it under the terms of the BSD 3-clause License. If software is modified to produce derivative works, such modified software should be clearly marked, so as not to confuse it with the version available from LANL. Full text of the BSD 3-clause License can be found in the License file in the main development branch of the repository.
#
##############################################################################
# BSD 3-clause license:
# Copyright 2015- Triad National Security, LLC
#
# Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##############################################################################
# POC: Stephan Eidenbenz, eidenben@lanl.gov
# Date: March 25, 2015
# Copyright: Open source, must acknowledge original author
# Purpose: PDES Engine in CPython and PyPy, mirroring most of the original LuaJIT version of Simian JIT-PDES
# NOTE: If speed rivaling C/C++ PDES engines is desired, consider adopting the LuaJIT version of Simian JIT-PDES
#
##############################################################################
# Changelog:
#
# NOTE: 5/9/2020: Changes: Nandakishore Santhi
# Simian for Python 3.7
# Combined all Simian modules into a single standalone module file
# Updated LICENSE and COPYRIGHT notices
#
##############################################################################
#
# An attempt to replicate the well known pHOLD PDES benchmark (with MPI)

from simian import Simian
import random, math, sys

opt = (str(sys.argv[1]).lower() == "true")
simName, startTime, endTime, minDelay, useMPI = "PHOLD", 0, 100, 1, True
simianEngine = Simian(simName, startTime, endTime, minDelay, useMPI, opt)

count = 64
lookahead = minDelay

def exponential(mean):
	return -math.log(random.random())/mean

class Node(simianEngine.Entity):
	def __init__(self, baseInfo, *args):
		super(Node, self).__init__(baseInfo)
		self.targetId = 0
		self.offset = 0

	def generate(self, *args):
		self.targetId = random.randrange(count)
		#self.offset = exponential(1) + lookahead
		self.offset = lookahead

		self.reqService(self.offset, "generate", None, "Node", self.targetId)

	def saveState(self):
		state = {"targetId" : self.targetId, "offset" : self.offset}
		return dict(state)

	def recoverState(self,state):
		self.targetId = state["targetId"]
		self.offset = state["offset"]
		return

for i in range(count):
	simianEngine.addEntity("Node", Node, i)

for i in range(count):
	simianEngine.schedService(0, "generate", None, "Node", i)

simianEngine.run()
simianEngine.exit()

File: SimianLua/Examples/phold-opt.lua
--[[
Copyright (c) 2015, Los Alamos National Security, LLC
All rights reserved.

Copyright 2015. Los Alamos National Security, LLC. This software was produced under U.S. Government contract DE-AC52-06NA25396 for Los Alamos National Laboratory (LANL), which is operated by Los Alamos National Security, LLC for the U.S. Department of Energy. The U.S. Government has rights to use, reproduce, and distribute this software.  NEITHER THE GOVERNMENT NOR LOS ALAMOS NATIONAL SECURITY, LLC MAKES ANY WARRANTY, EXPRESS OR IMPLIED, OR ASSUMES ANY LIABILITY FOR THE USE OF THIS SOFTWARE.  If software is modified to produce derivative works, such modified software should be clearly marked, so as not to confuse it with the version available from LANL.

Additionally, redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
    Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer. 
    Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution. 
    Neither the name of Los Alamos National Security, LLC, Los Alamos National Laboratory, LANL, the U.S. Government, nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission. 
THIS SOFTWARE IS PROVIDED BY LOS ALAMOS NATIONAL SECURITY, LLC AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL LOS ALAMOS NATIONAL SECURITY, LLC OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
]]

--[[
--Author: Nandakishore Santhi
--Date: 23 November, 2014
--Copyright: Open source, must acknowledge original author
--Purpose: JITed PDES Engine in LuaJIT
--  Simple example simulation script
--]]
package.path = "Simian/?.lua;" .. package.path

local Simian = require "simian"
local ln, random = math.log, math.random

local opt = (tostring(arg[1]):lower() == "true")
local simName, startTime, endTime, minDelay, useMPI = "PHOLD", 0, 100, 1, true

local count = 64 
local lookahead = minDelay

local function exponential(lambda)
    return -ln(random())/lambda
end

local Node = Simian.Entity("Node")
local targetId = 0
local offset = 0

function Node:generate(...)
    targetId = random(0, count-1)
    --targetId = (self.engine.rank + 1) % self.engine.size
    --targetId = self.engine.rank

    --offset = exponential(1) + lookahead
    offset = lookahead

    --self.out:write("Time "
    --            .. self.engine.now
    --            .. ": Waking " .. self.targetId
    --            .. " at " .. self.offset .. " from now\n")

    --print(self.engine.rank, " Execute event to ", targetId)
    self:reqService(offset, "generate", nil, "Node", targetId)
end

function Node:saveState()
    local state = {tid = targetId, off = offset,}
    return state
end

function Node:recoverState(state)
    targetId = state.tid
    offset = state.off
end

--Initialize Simian
Simian:init(simName, startTime, endTime, minDelay, useMPI, opt)

for i=0,count-1 do
    Simian:addEntity("Node", Node, i)
end

for i=0,count-1 do
    Simian:schedService(0, "generate", nil, "Node", i)
end

Simian:run()
Simian:exit()
